packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/launcher/errors/DevLauncherUncaughtExceptionHandler.kt

```diff
@@ -0,0 +1,85 @@
+package expo.modules.devlauncher.launcher.errors
+
+import android.app.Activity
+import android.app.Application
+import android.os.Bundle
+import android.os.Process
+import android.util.Log
+import java.lang.ref.WeakReference
+import java.util.*
+import kotlin.concurrent.schedule
+import kotlin.system.exitProcess
+
+class DevLauncherUncaughtExceptionHandler(
+  application: Application,
+  private val defaultUncaughtHandler: Thread.UncaughtExceptionHandler?
+) : Thread.UncaughtExceptionHandler {
+  private val applicationHolder = WeakReference(application)
+  private var exceptionWasReported = false
+  private var timerTask: TimerTask? = null
+
+  init {
+    application.registerActivityLifecycleCallbacks(object : Application.ActivityLifecycleCallbacks {
+      override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
+        if (exceptionWasReported && activity is DevLauncherErrorActivity) {
+          timerTask?.cancel()
+          timerTask = null
+          exceptionWasReported = false
+          return
+        }
+      }
+
+      override fun onActivityStarted(activity: Activity) = Unit
+
+      override fun onActivityResumed(activity: Activity) = Unit
+
+      override fun onActivityPaused(activity: Activity) = Unit
+
+      override fun onActivityStopped(activity: Activity) = Unit
+
+      override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) = Unit
+
+      override fun onActivityDestroyed(activity: Activity) = Unit
+    })
+
+  }
+
+  override fun uncaughtException(thread: Thread, exception: Throwable) {
+    // The same exception can be reported multiple times.
+    // We handle only the first one.
+    if (exceptionWasReported) {
+      return
+    }
+
+    exceptionWasReported = true
+    Log.e("DevLauncher", "DevLauncher tries to handle uncaught exception.", exception)
+    applicationHolder.get()?.let {
+      DevLauncherErrorActivity.showFatalError(
+        it,
+        DevLauncherAppError(exception.message, exception)
+      )
+    }
+
+    // We don't know if the error screen will show up.
+    // That's why we schedule a simple function which will check
+    // if the error was handle properly or will fallback
+    // to the default exception handler.
+    timerTask = Timer().schedule(2000) {
```

Comment:
In what situations would the error screen not show up? When we're running an app in development vs. production mode? Would prefer to have something more predictable than a timer for determining whether or not to call the default handler.

Reply:
Ok, so firstly, the development client works only in development mode. It won't affect the production builds. And yes, the error screen won't show up in that case. I would also prefer something different than a timer, but I couldn't find another solution. Unofrtunelty, I'm not even sure if that is possible. In that case, we have to deal with the weird state of the application and the system doesn't allow us to open another activity.

Reply:
Sorry, I meant a development vs. production JS bundle, not debug vs. release build. But what I am really trying to ask is, what are the cases in which we want the error screen to show up but it won't?

Reply:
If the crash occurs very early. For example, if you throw an exception in the `onCreate` method in `MainApplication` or the same method in Activity, the error scene won't show up. I think it is because the Android context isn't set up, in those cases and we try to use it to open another activity. 

The js mode isn't respected at all here. We always will try to open the error screen. I think it's the desired behavior.

Reply:
I see. If there's no other way to check this, then I think this is ok but would be good to be explicit in the comments about the situations the timer is meant for.